### PR TITLE
feat: enhance MCP tool response handling to support image responses

### PR DESCRIPTION
**Pull request by shivangag**
### Related GitHub Issue

I couldn't find an approved issue related to this. 

### Description

This PR enhances MCP (Model Context Protocol) tool response handling to support images alongside text content. The implementation allows MCP servers to return images in their tool responses, which are then properly displayed in the Roo Code UI.

**Key implementation details:**
- Modified [`processToolContent()`](src/core/tools/useMcpToolTool.ts) to extract and return both text and images from MCP tool results
- Updated [`executeToolAndProcessResult()`](src/core/tools/useMcpToolTool.ts) to handle image data and pass it through the response pipeline
- Enhanced [`combineCommandSequences()`](src/shared/combineCommandSequences.ts) to preserve images from MCP server responses during command sequence processing
- Updated UI components ([`McpExecution.tsx`](webview-ui/src/components/chat/McpExecution.tsx) and [`ChatRow.tsx`](webview-ui/src/components/chat/ChatRow.tsx)) to render images alongside text responses

**Design choices:**
- Images are handled as part of the tool result structure, maintaining backward compatibility with text-only responses
- The implementation supports both text-only, image-only, and combined text+image responses from MCP tools

### Test Procedure

**Unit Tests Added:**
- Added comprehensive tests in [`useMcpToolTool.spec.ts`](src/core/tools/__tests__/useMcpToolTool.spec.ts) to verify:
  - Correct extraction of both text and images from tool results
  - Proper handling of image-only responses
  - Backward compatibility with text-only responses
- Updated tests in [`combineCommandSequences.spec.ts`](src/shared/__tests__/combineCommandSequences.spec.ts) to ensure images are preserved during command sequence processing

**Manual Testing:**
1. Connect to an MCP server that supports image responses (e.g., browser MCP server with screenshot capability)
2. Execute a tool that returns images (e.g., take a screenshot)
3. Verify that both text and images are displayed correctly in the chat interface
4. Test with various MCP tools to ensure backward compatibility with text-only responses

**Reviewers can test by:**
1. Running the test suite: `npm test -- useMcpToolTool.spec.ts`
2. Setting up an MCP server with image capabilities and testing the integration manually

### Pre-Submission Checklist

- [ ] **Issue Linked**: This PR is linked to an approved GitHub Issue (see "Related GitHub Issue" above).
- [x] **Scope**: My changes are focused on the linked issue (one major feature/fix per PR).
- [x] **Self-Review**: I have performed a thorough self-review of my code.
- [x] **Testing**: New and/or updated tests have been added to cover my changes (if applicable).
- [ ] **Documentation Impact**: I have considered if my changes require documentation updates (see "Documentation Updates" section below).
- [x] **Contribution Guidelines**: I have read and agree to the [Contributor Guidelines](/CONTRIBUTING.md).

### Screenshots / Videos

**Before:** MCP tool responses only displayed text content, images were not supported.
Screenshot 2025-06-27 at 2 32 26 PM

**After:** MCP tool responses now display both text and images seamlessly in the chat interface.
`codicon-file-media` indicates if images are present in mcp response and the number of images.
Screenshot 2025-06-27 at 2 18 30 PM
Images are presented using the `Thumbnail` component inside `mcp-server-response` block.
Screenshot 2025-06-27 at 2 19 29 PM

### Documentation Updates

- [ ] No documentation updates are required.
- [x] Yes, documentation updates are required. The MCP integration documentation should be updated to mention image support capabilities for tool responses.

### Additional Notes

- This enhancement maintains full backward compatibility with existing MCP servers that only return text
- The implementation follows the existing patterns for handling MCP tool responses
- Images are processed and displayed using the same infrastructure as other image content in Roo Code

### Improvements
Old chat loaded from memory show tool call argument inside the response.
Screenshot 2025-06-27 at 2 19 44 PM

### CODE REVIEW

File: src/core/tools/useMcpToolTool.ts [hunk 1]
```diff
@@ -89,24 +89,39 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
+	const images: string[] = []
+
+	toolResult.content.forEach((item: any) => {
+		if (item.type === "text") {
+			textParts.push(item.text)
+		} else if (item.type === "image") {
+			if (item.data && item.mimeType) {
+				const validImageTypes = ["image/png", "image/jpeg", "image/gif", "image/webp"]
+				if (validImageTypes.includes(item.mimeType)) {
+					const dataUrl = `data:${item.mimeType};base64,${item.data}`
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts]: Have you considered adding validation for maximum image size? Large base64-encoded images could potentially cause memory issues. Maybe we could add a configurable limit or at least log a warning for images above a certain threshold (e.g., 10MB)?

File: src/core/tools/useMcpToolTool.ts [hunk 2]
```diff
@@ -89,24 +89,39 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
+					images.push(dataUrl)
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts]: What happens if the base64 data is corrupted or invalid? Would it be worth wrapping this in a try-catch to handle potential errors when constructing the data URL? This could prevent the entire tool response from failing due to a single corrupted image.

File: src/core/tools/useMcpToolTool.ts [hunk 3]
```diff
@@ -89,24 +89,39 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
-				return item.text
+	const textParts: string[] = []
+	const images: string[] = []
+
+	toolResult.content.forEach((item: any) => {
+		if (item.type === "text") {
+			textParts.push(item.text)
+		} else if (item.type === "image") {
+			if (item.data && item.mimeType) {
+				const validImageTypes = ["image/png", "image/jpeg", "image/gif", "image/webp"]
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts]: Would it make sense to extract this array as a constant at the module level? Something like `const SUPPORTED_IMAGE_TYPES = ["image/png", "image/jpeg", "image/gif", "image/webp"]`. This would make it easier to maintain and potentially reuse elsewhere.

File: src/core/tools/__tests__/useMcpToolTool.spec.ts
Comment (daniel-lxs) [inline src/core/tools/__tests__/useMcpToolTool.spec.ts:1]: The test coverage looks good! Have you considered adding a few edge case tests? For example:
- Testing with corrupted base64 data
- Testing with very large images to ensure performance
- Testing with unsupported MIME types (I see the warning is logged, but a test would be good)
- Testing with malformed image content (missing data or mimeType)

File: src/core/tools/useMcpToolTool.ts [hunk 4]
```diff
@@ -89,24 +89,39 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
-		return ""
+		return { text: "", images: [] }
 	}
 
-	return toolResult.content
-		.map((item: any) => {
-			if (item.type === "text") {
-				return item.text
+	const textParts: string[] = []
+	const images: string[] = []
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts]: Should we consider implementing a limit on the number of images that can be returned in a single response? Without a limit, a malicious or buggy MCP server could potentially return hundreds of images, causing performance issues. Maybe add a configurable maximum (e.g., 10-20 images) and log a warning if exceeded?

**Review by daniel-lxs — COMMENTED** [review]
Hey @shivangag,

I noticed this PR doesn't have a linked GitHub issue. As per the contributing guidelines, all PRs need to be linked to an approved issue. Could you create an issue describing this enhancement and link it to this PR? This helps us keep things traceable and ensures features are properly discussed before implementation.

I'll mark this PR as "In progress" for now until the dev team has a chance to review the issue.

**Comment by shivangag** [comment]
Thank you @daniel-lxs for the thorough review and excellent suggestions! I've implemented all the recommended enhancements to make MCP image handling more robust and secure.

#### Security & Performance Controls
- **Image size limits**: Added configurable `mcpMaxImageSizeMB` setting with 10MB default
- **Count limits**: Added configurable `mcpMaxImagesPerResponse` setting with 20 image default
- **User controls**: Both settings are configurable through the MCP settings panel

#### Robustness Improvements  
- **Error handling**: Wrapped image processing in try-catch blocks to prevent crashes from corrupted data
- **Validation**: Added base64 regex validation and type checking for data integrity
- **Graceful degradation**: Invalid images are skipped with detailed warning logs instead of failing

#### Code Quality Enhancements
- **Maintainability**: Extracted `SUPPORTED_IMAGE_TYPES` as a module-level constant
- **Test coverage**: Added comprehensive test suite covering all suggested edge cases including corrupted data, large images, unsupported MIME types, and malformed content

The implementation now safely handles potentially malicious or buggy MCP servers while maintaining full backward compatibility with existing functionality. I believe this addresses all the security, performance, and reliability concerns raised in the review.

Also created GitHub issue #5233 to properly document this feature enhancement.

File: webview-ui/src/components/mcp/McpView.tsx [hunk 5]
```diff
@@ -105,6 +110,52 @@ const McpView = ({ onDone }: McpViewProps) => {
...
+							<div style={{ marginBottom: 10 }}>
+								<VSCodeTextField
+									value={mcpMaxImagesPerResponse.toString()}
+									onChange={(e: any) => {
+										const value = parseInt(e.target.value) || 20
+										setMcpMaxImagesPerResponse(value)
+										vscode.postMessage({ type: "mcpMaxImagesPerResponse", value })
+									}}
+									style={{ width: "100px" }}>
+									Max Images Per Response
```
Comment (daniel-lxs) [inline webview-ui/src/components/mcp/McpView.tsx]: The label "Max Images Per Response" should be translated. Could we add a translation key like `mcp:imageSettings.maxImagesLabel` to the mcp.json translation files?

Similarly, the description text on lines 131-132 should use a translation key like `mcp:imageSettings.maxImagesDescription`.

File: webview-ui/src/components/mcp/McpView.tsx [hunk 6]
```diff
@@ -105,6 +110,52 @@ const McpView = ({ onDone }: McpViewProps) => {
...
+							<div style={{ marginBottom: 10 }}>
+								<VSCodeTextField
+									value={mcpMaxImageSizeMB.toString()}
+									onChange={(e: any) => {
+										const value = parseFloat(e.target.value) || 10
+										setMcpMaxImageSizeMB(value)
+										vscode.postMessage({ type: "mcpMaxImageSizeMB", value })
+									}}
+									style={{ width: "100px" }}>
+									Max Image Size (MB)
```
Comment (daniel-lxs) [inline webview-ui/src/components/mcp/McpView.tsx]: The label "Max Image Size (MB)" should also be translated. Could we use a translation key like `mcp:imageSettings.maxSizeLabel`?

And the description on lines 153-154 should use a key like `mcp:imageSettings.maxSizeDescription`.

File: webview-ui/src/components/mcp/McpView.tsx [hunk 7]
```diff
@@ -105,6 +110,52 @@ const McpView = ({ onDone }: McpViewProps) => {
 							</div>
 						</div>
 
+						<div style={{ marginBottom: 15 }}>
+							<div style={{ marginBottom: 10 }}>
+								<VSCodeTextField
+									value={mcpMaxImagesPerResponse.toString()}
```
Comment (daniel-lxs) [inline webview-ui/src/components/mcp/McpView.tsx:120]: Could we add input validation to these number fields? Consider adding:
```tsx
type="number"
min="1"
max="100"
```
This would prevent invalid inputs and improve the user experience. Also, the current `parseInt(e.target.value) || 20` might reset to 20 if the user types 0, which could be confusing.

**Review by daniel-lxs — CHANGES_REQUESTED** [review]
Hey @shivangag, I left some new suggestions from my review. Let me know if you have any questions!

File: webview-ui/src/components/mcp/McpView.tsx (hunk 7 above)
Comment (shivangag) [reply webview-ui/src/components/mcp/McpView.tsx:120]: @daniel-lxs There seems to be no prop `type` in `VSCodeTextField`. I've added input validation in `onChange` handler and an error state for now.
We can also use `Slider` in place of `VSCodeTextField`. Let me know what you think?

File: src/core/tools/useMcpToolTool.ts [hunk 8]
```diff
@@ -89,24 +91,87 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
+
+						// Basic validation for base64 format
+						const base64Regex = /^[A-Za-z0-9+/]*={0,2}$/
+						if (!base64Regex.test(item.data.replace(/\s/g, ""))) {
+							console.warn("Invalid MCP ImageContent: base64 data contains invalid characters")
+							return
+						}
+
+						// Check image size
+						const imageSizeMB = calculateImageSizeMB(item.data)
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts]: The image size calculation happens after the entire base64 string is already in memory. For very large images, this could cause memory spikes before the size check rejects them.

Consider adding an early size check based on string length:
```typescript
// Quick size check before full validation
const approximateSizeMB = (item.data.length * 0.75) / (1024 * 1024);
if (approximateSizeMB > maxImageSizeMB * 1.5) { // Allow some margin for encoding overhead
    console.warn(`MCP image likely exceeds size limit based on string length`);
    return;
}
```

This would prevent processing extremely large strings that are obviously over the limit.

File: webview-ui/src/components/common/Thumbnails.tsx [hunk 9]
```diff
@@ -52,18 +60,43 @@ const Thumbnails = ({ images, style, setImages, onHeightChange }: ThumbnailsProp
...
+								borderRadius: 4,
+								backgroundColor: "var(--vscode-input-background)",
+								border: "1px solid var(--vscode-input-border)",
+								display: "flex",
+								justifyContent: "center",
+								alignItems: "center",
+								cursor: "pointer",
+							}}
+							onClick={() => handleImageClick(image)}
+							title="Failed to load image">
```
Comment (daniel-lxs) [inline webview-ui/src/components/common/Thumbnails.tsx]: This hardcoded English string needs to be translated. Please add a translation key for "Failed to load image".

Suggested fix:
```typescript
title={t("common:thumbnails.failedToLoad")}
```

Don't forget to add the translation to all locale files.

File: webview-ui/src/components/common/Thumbnails.tsx [hunk 10]
```diff
@@ -52,18 +60,43 @@ const Thumbnails = ({ images, style, setImages, onHeightChange }: ThumbnailsProp
...
+								className="codicon codicon-file-media"
+								style={{
+									color: "var(--vscode-descriptionForeground)",
+									fontSize: 16,
+								}}></span>
+						</div>
+					) : (
+						<img
+							src={image}
+							alt={`Thumbnail ${index + 1}`}
```
Comment (daniel-lxs) [inline webview-ui/src/components/common/Thumbnails.tsx]: The alt text contains a hardcoded English string with string interpolation. This needs to be translated.

Suggested fix:
```typescript
alt={t("common:thumbnails.altText", { index: index + 1 })}
```

The translation key should support interpolation, e.g., "Thumbnail {{index}}"

File: src/core/tools/useMcpToolTool.ts [hunk 11]
```diff
@@ -89,24 +91,87 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
-	return toolResult.content
-		.map((item: any) => {
-			if (item.type === "text") {
-				return item.text
+	const textParts: string[] = []
+	const images: string[] = []
+
+	// Get MCP settings from the extension's global state
+	const state = await cline.providerRef.deref()?.getState()
+	const maxImagesPerResponse = state?.mcpMaxImagesPerResponse ?? 20
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts]: The current implementation uses the settings values without validating whether they're within reasonable bounds. For example, if someone manually edits their settings file and sets a negative number or an extremely large value, it could lead to unexpected behavior.

It might be worth adding validation like this:

```ts
const maxImagesPerResponse = Math.max(1, Math.min(100, state?.mcpMaxImagesPerResponse ?? 20));
const maxImageSizeMB = Math.max(0.1, Math.min(50, state?.mcpMaxImageSizeMB ?? 10));
````

This helps keep the values within safe operational limits, even in cases where the settings file is corrupted or manually modified.

File: webview-ui/src/components/chat/McpExecution.tsx [hunk 12]
```diff
@@ -212,15 +215,23 @@ export const McpExecution = ({
...
-						{responseText && responseText.length > 0 && (
-							<Button variant="ghost" size="icon" onClick={onToggleResponseExpand}>
-								<ChevronDown
-									className={cn("size-4 transition-transform duration-300", {
-										"rotate-180": isResponseExpanded,
-									})}
-								/>
-							</Button>
-						)}
+						{(responseText && responseText.length > 0) || (images && images.length > 0) ? (
```
Comment (daniel-lxs) [inline webview-ui/src/components/chat/McpExecution.tsx:218]: The image count indicator is shown even when the response is collapsed. Is this intentional? It might be confusing to show a count for content that isn't visible.

Maybe you could:
1. Only showing the count when expanded
2. Adding a tooltip to explain what the count represents
3. Using different styling to indicate collapsed content with images

File: src/core/tools/useMcpToolTool.ts [hunk 13]
```diff
@@ -89,24 +91,87 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
+	const maxImageSizeMB = state?.mcpMaxImageSizeMB ?? 10
+
+	toolResult.content.forEach((item: any) => {
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts:289]: Images are validated sequentially in the forEach loop. For responses with many images, this could be slow.

Consider batching validation or using Promise.all for parallel processing:
```typescript
const validatedImages = await Promise.all(
  toolResult.content
    .filter((item: any) => item.type === "image")
    .map(async (item: any) => validateAndProcessImage(item))
);
images.push(...validatedImages.filter(Boolean));
```

This would improve performance for MCP tools that return multiple images.

File: src/core/tools/useMcpToolTool.ts [hunk 14]
```diff
@@ -89,24 +91,87 @@ async function sendExecutionStatus(cline: Task, status: McpExecutionStatus): Pro
...
+				if (SUPPORTED_IMAGE_TYPES.includes(item.mimeType)) {
+					try {
+						// Validate base64 data before constructing data URL
+						if (typeof item.data !== "string" || item.data.trim() === "") {
+							console.warn("Invalid MCP ImageContent: base64 data is not a valid string")
+							return
+						}
+
+						// Basic validation for base64 format
+						const base64Regex = /^[A-Za-z0-9+/]*={0,2}$/
```
Comment (daniel-lxs) [inline src/core/tools/useMcpToolTool.ts]: The base64 validation regex could be extracted as a constant for better maintainability and potential reuse:

```typescript
const BASE64_REGEX = /^[A-Za-z0-9+/]*={0,2}$/;
```

This would make it easier to maintain and test the validation pattern.

**Comment by mahmoudmoravej** [comment]
Hi, is there any update on this issue? Without it, we’re unable to use tools like the Figma MCP server, which depend heavily on the screenshots retrieved through its `get_image` tool.

File: webview-ui/src/components/chat/McpExecution.tsx (hunk 12 above)
Comment (shivangag) [reply webview-ui/src/components/chat/McpExecution.tsx:218]: Idea was to easily glance at the number of images in response without toggling. I've added a tooltip.

File: src/core/tools/useMcpToolTool.ts (hunk 8 above)
Comment (shivangag) [reply src/core/tools/useMcpToolTool.ts]: Done.

File: src/core/tools/useMcpToolTool.ts (hunk 11 above)
Comment (shivangag) [reply src/core/tools/useMcpToolTool.ts]: Done.

File: webview-ui/src/components/common/Thumbnails.tsx (hunk 10 above)
Comment (shivangag) [reply webview-ui/src/components/common/Thumbnails.tsx]: Added.

File: webview-ui/src/components/common/Thumbnails.tsx (hunk 9 above)
Comment (shivangag) [reply webview-ui/src/components/common/Thumbnails.tsx]: Added.

File: src/core/tools/useMcpToolTool.ts (hunk 13 above)
Comment (shivangag) [reply src/core/tools/useMcpToolTool.ts:289]: Done.

**Comment by shivangag** [comment]
@daniel-lxs please review!

**Comment by daniel-lxs** [comment]
This PR seems to have conflicts, but I'm not sure how many

**Comment by hannesrudolph** [comment]
> This PR seems to have conflicts, but I'm not sure how many

Maybe we get merge resolver on this?

**Comment by shivangag** [comment]
> This PR seems to have conflicts, but I'm not sure how many

@daniel-lxs I have resolved the conflicts after rebasing from main.

**Comment by daniel-lxs** [comment]
We are currently trying to get base64-string images out of the webview. They are a big contributor to the grey screen issue we have seen in some situations. You can see an example of this here: https://github.com/RooCodeInc/Roo-Code/pull/8225.

I would think that this PR will also benefit from a change like that. My recommendation is that we wait for PR #8225 to be merged, and then you can probably recycle some of the functionality to apply it here as well.

**Comment by hannesrudolph** [comment]
Closing because the linked issue is being closed.
